File: ex1/siso_ser.py
import numpy as np
import matplotlib.pyplot as plt

# Define QPSK constellation points
QPSK_vec = np.array([1+1j, 1-1j, -1+1j, -1-1j]) / np.sqrt(2)

# SNR in dB and linear scale
SNRdB = np.arange(0, 31, 2)
SNR = 10 ** (SNRdB / 10)

# Initialize SER array
SER_rayligh = np.zeros(len(SNRdB))
SER_awgn = np.zeros(len(SNRdB))
N = 500000  # Number of symbols

for is_rayligh in [False, True]:
    # SER calculation for each SNR value
    for k in range(len(SNRdB)):
        rho = 1 / np.sqrt(SNR[k])

        # Signal Generation
        s = np.random.choice(QPSK_vec, N)
        if is_rayligh:
            h = (np.random.randn(N) + 1j * np.random.randn(N)) / np.sqrt(2)
        else:
            h = np.ones(N)
        
        n = (np.random.randn(N) + 1j * np.random.randn(N)) / np.sqrt(2)
        y = h * s + rho * n

        # Signal Detection
        s_hat = y / h
        s_tilde = np.sign(np.real(s_hat)) / np.sqrt(2) + 1j * np.sign(np.imag(s_hat)) / np.sqrt(2)

        # Error calculation
        NumOfErrors = np.sum(np.abs(s_tilde - s) > np.finfo(float).eps)
        if is_rayligh:
            SER_rayligh[k] = NumOfErrors / N
        else:
            SER_awgn[k] = NumOfErrors / N

# Plotting
plt.semilogy(SNRdB, SER_rayligh, marker='x', label='Rayleigh')
plt.semilogy(SNRdB, SER_awgn, marker='o', label='AWGN')
plt.grid(True, which='both')
plt.xlabel('SNR (dB)')
plt.ylabel('SER')
plt.title('SISO Symbol Error Rate - Rayleigh and AWGN channels')
plt.legend()
plt.show()

File: ex1/SISO_SER.m
close all; clear; clc

QPSK_vec=[1+1i 1-1i -1+1i -1-1i]/sqrt(2);
SNRdB=0:2:30;
SNR=10.^(SNRdB./10);

SER=zeros(size(SNRdB));
N=500000;

for k=1:length(SNRdB),
    rho=1./sqrt(SNR(k));
    % Signal Generation
    s=randsrc(N,1,QPSK_vec);
    h=(randn(N,1)+1i*randn(N,1))/sqrt(2);
    n=(randn(N,1)+1i*randn(N,1))/sqrt(2);
    y=h.*s+rho*n;
    
    % Signal Detection    
    s_hat=y./h;
    s_tilde=sign(real(s_hat))/sqrt(2)+1i*sign(imag(s_hat))/sqrt(2);
    
    NumOfErrors=sum(abs(s_tilde-s)>eps); 
    SER(k)=NumOfErrors/N; 
end;

semilogy(SNRdB,SER); grid; 
xlabel('SNR(dB)'); 
ylabel('SER'); 
title('SISO Rayleigh'); 